go-controller/pkg/ovn/ovn.go

```diff
@@ -220,34 +244,45 @@ func getPodNamespacedName(pod *kapi.Pod) string {
 	return util.GetLogicalPortName(pod.Namespace, pod.Name)
 }
 
-// NewOvnController creates a new OVN controller for creating logical network
-// infrastructure and policy
-func NewOvnController(ovnClient *util.OVNClientset, wf *factory.WatchFactory, stopChan <-chan struct{}, addressSetFactory addressset.AddressSetFactory,
-	libovsdbOvnNBClient libovsdbclient.Client, libovsdbOvnSBClient libovsdbclient.Client,
-	recorder record.EventRecorder) *Controller {
+// NewControllerInfo creates controllerInfo of the controller
+func NewControllerInfo(client clientset.Interface, kube kube.Interface, wf *factory.WatchFactory,
+	recorder record.EventRecorder, nbClient libovsdbclient.Client,
+	sbClient libovsdbclient.Client, podRecorder *metrics.PodRecorder, SCTPSupport bool) controllerInfo {
+	return controllerInfo{
+		client:       client,
+		kube:         kube,
+		watchFactory: wf,
+		recorder:     recorder,
+		nbClient:     nbClient,
+		sbClient:     sbClient,
+		podRecorder:  podRecorder,
+		SCTPSupport:  SCTPSupport,
+	}
+}
+
+// NewDefaultL3Controller creates a new OVN controller for creating logical network
+// infrastructure and policy for default l3 network
+func NewDefaultL3Controller(cInfo controllerInfo,
+	defaultStopChan chan struct{}, defaultWg *sync.WaitGroup,
```

Comment:
The stop chan and the wait group, and anything that is related to managing the lifetime of the controller should be included in the interface that the controller manager uses to handle all the controllers. In general, this is the purpose of that interface, so that it is clear how the lifetime of the managed controllers is handled.
Example:
```
type ManagedNetworkController interface {
        Init() error
        Run(ctx context.Context, stopchan <-chan struct{}, doneWg *sync.WaitGroup) error
	Stop()
}
```

Also wanted to add that a context and a stopchan might be used for the same purpose so it might be confusing to have them both. In that case we should document exaclty how each one is being used.

Reply:
I do't see needs for this, today oc.Run() does not take stopChan.

Reply:
It does not take it but it uses it. The reason it does not take it is because it did not really matter much before because there was no interface. But now that we are introducing an interface, the interface should clearly depict how the network controllers are started or stopped. This means that we cannot pass things that are involved in this in the constructor as the constructor is not part of the interface.

Reply:
I don't have strong options on one way or another, but I don't think I understand your concern. stopChan and wg is now perController, why there is a need to take them as argument of the interface as there should not be any confusions.

Reply:
Sorry, I missed that  stopChan and wg are per-controller.